pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -0,0 +1,309 @@
+		return nil, err
+	}
+
+	if len(tenantIDs) <= 1 {
+		return m.upstream.Querier(ctx, mint, maxt)
+	}
+
+	var queriers = make([]storage.Querier, len(tenantIDs))
+	for pos, tenantID := range tenantIDs {
+		q, err := m.upstream.Querier(
+			user.InjectOrgID(ctx, tenantID),
+			mint,
+			maxt,
+		)
+		if err != nil {
+			return nil, err
+		}
+		queriers[pos] = q
+	}
+
+	return &mergeQuerier{
+		queriers:        queriers,
+		tenantIDs:       tenantIDs,
+		tenantLabelName: defaultTenantLabel,
+	}, nil
+}
+
+const retainExistingPrefix = "original_"
+
+// mergeQuerier aggregates the results from underlying queriers and adds a
+// label tenantLabelName to identify the tenant ID that the metric resulted
+// from.
+// If the label tenantLabelName is already existing, its value is
+// overwritten by the tenant ID and the previous value is exposed through a new
+// label prefixed with "original_". This behaviour is not implemented recursively
+type mergeQuerier struct {
+	queriers        []storage.Querier
+	tenantIDs       []string
+	tenantLabelName model.LabelName
+}
+
+// LabelValues returns all potential values for a label name.
+// It is not safe to use the strings beyond the lifefime of the querier.
+// For the label "tenantLabelName" it will return all the tenant IDs available.
+// For the label "original_" + tenantLabelName it will return all the values
+// of the underlying queriers for tenantLabelName.
+func (m *mergeQuerier) LabelValues(name string) ([]string, storage.Warnings, error) {
+	if name == string(m.tenantLabelName) {
+		return m.tenantIDs, nil, nil
+	}
+
+	// ensure the name of a retained tenant id label gets handled under the
+	// origignal label name
+	if name == retainExistingPrefix+string(m.tenantLabelName) {
+		name = string(m.tenantLabelName)
+	}
+
+	return m.mergeDistinctStringSlice(func(q storage.Querier) ([]string, storage.Warnings, error) {
+		return q.LabelValues(name)
+	})
+}
+
+// LabelNames returns all the unique label names present in the underlying
+// queriers. It also add the tenantLabelName and if present in the original
+// results the "original_" tenantLabelName.
+func (m *mergeQuerier) LabelNames() ([]string, storage.Warnings, error) {
+	labelNames, warnings, err := m.mergeDistinctStringSlice(func(q storage.Querier) ([]string, storage.Warnings, error) {
+		return q.LabelNames()
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// check if the tenant label exists in the original result
+	var tenantLabelExists bool
+	for _, labelName := range labelNames {
+		if labelName == string(m.tenantLabelName) {
+			tenantLabelExists = true
+			break
+		}
+	}
+
+	if tenantLabelExists {
+		// the tenant label exists in the original result, so it had been retained
+		labelNames = append(labelNames, retainExistingPrefix+string(m.tenantLabelName))
+	} else {
+		// the tenant label was not existing, so it need to get added
+		labelNames = append(labelNames, string(m.tenantLabelName))
+	}
+	sort.Strings(labelNames)
```

Comment:
Since you're just appending a single string, it would be more efficient to append it in the right place (look for it using `sort.SearchStrings()`) instead of appending to the end and then resorting it entirely.